packages/interaction-tracking/src/InteractionTracking.js

```diff
@@ -0,0 +1,341 @@
+
+  // A batch of work has started for a set of interactions.
+  // When this work is complete, onWorkStopped will be called.
+  // Work is not always completed synchronously; yielding may occur in between.
+  // A batch of async/yieldy work may also be re-started before completing.
+  // In that case, onWorkStarted may be called more than once before onWorkStopped.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStarted: (interactions: Set<Interaction>, threadID: number) => void,
+
+  // A batch of work has completed for a set of interactions.
+  // Work is done by a "thread" which is identified by a unique ID.
+  onWorkStopped: (interactions: Set<Interaction>, threadID: number) => void,
+};
+
+export type InteractionsRef = {
+  current: Set<Interaction>,
+};
+
+export type SubscriberRef = {
+  current: Subscriber | null,
+};
+
+const DEFAULT_THREAD_ID = 0;
+
+// Counters used to generate unique IDs.
+let interactionIDCounter: number = 0;
+let threadIDCounter: number = 0;
+
+// Set of currently tracked interactions.
+// Interactions "stack"–
+// Meaning that newly tracked interactions are appended to the previously active set.
+// When an interaction goes out of scope, the previous set (if any) is restored.
+let interactionsRef: InteractionsRef | null = null;
+
+// Listener(s) to notify when interactions begin and end.
+// Note that subscribers are only supported when enableInteractionTrackingObserver is enabled.
+let subscriberRef: SubscriberRef | null = null;
+
+if (enableInteractionTracking) {
+  interactionsRef = {
+    current: new Set(),
+  };
+  if (enableInteractionTrackingObserver) {
+    subscriberRef = {
+      current: null,
+    };
+  }
+}
+
+// These values are exported for libraries with advanced use cases (i.e. React).
+// They should not typically be accessed directly.
+export {interactionsRef as __interactionsRef, subscriberRef as __subscriberRef};
+
+export function clear(callback: Function): any {
+  if (!enableInteractionTracking) {
+    return callback();
+  }
+
+  const prevInteractions = ((interactionsRef: any): InteractionsRef).current;
+  ((interactionsRef: any): InteractionsRef).current = new Set();
+
+  try {
+    return callback();
+  } finally {
+    ((interactionsRef: any): InteractionsRef).current = prevInteractions;
+  }
+}
+
+export function getCurrent(): Set<Interaction> | null {
+  if (!enableInteractionTracking) {
+    return null;
+  } else {
+    return ((interactionsRef: any): InteractionsRef).current;
+  }
+}
+
+export function getThreadID(): number {
+  return ++threadIDCounter;
+}
+
+export function subscribe(subscriber: Subscriber): void {
+  if (enableInteractionTracking && enableInteractionTrackingObserver) {
+    invariant(
+      ((subscriberRef: any): SubscriberRef).current === null ||
+        ((subscriberRef: any): SubscriberRef).current === subscriber,
+      'Only one interactions subscriber may be registered at a time.',
+    );
```

Comment:
If only one can be scheduled at a time, maybe the API should be to mutate `current` directly.

Reply:
Oh that is what React does. Why does this other API exist at all?

Reply:
React mutates the `interactionRef` directly. Non-react things (e.g. Web Speed's code) will subscribe. I don't really want to expose the ref to things we don't directly control.